src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -211,57 +210,75 @@ public ProgrammingExerciseService(ProgrammingExerciseRepository programmingExerc
      * @throws GitAPIException If something during the communication with the remote Git repository went wrong
      * @throws IOException     If the template files couldn't be read
      */
-    @Transactional // TODO: apply the transaction on a smaller scope
-    // ok because we create many objects in a rather complex way and need a rollback in case of exceptions
     public ProgrammingExercise createProgrammingExercise(ProgrammingExercise programmingExercise, boolean isImportedFromFile) throws GitAPIException, IOException {
-        programmingExercise.generateAndSetProjectKey();
         final User exerciseCreator = userRepository.getUser();
-
         VersionControlService versionControl = versionControlService.orElseThrow();
+
+        programmingExercise.setSolutionParticipation(null);
+        programmingExercise.setTemplateParticipation(null);
+        programmingExerciseRepository.saveAndFlush(programmingExercise);
+
+        // Step 1: Setting constant facts for a programming exercise
+        programmingExercise.generateAndSetProjectKey();
         programmingExercise.setBranch(versionControl.getDefaultBranchOfArtemis());
+
+        // Step 2: Creating repositories for new exercise
         programmingExerciseRepositoryService.createRepositoriesForNewExercise(programmingExercise);
+        // Step 3: Initializing solution and template participation
         initParticipations(programmingExercise);
+
+        // Step 4a: Setting build plan IDs and URLs for template and solution participation
         setURLsAndBuildPlanIDsForNewExercise(programmingExercise);
 
-        // Save participations to get the ids required for the webhooks
+        // Step 4b: Connecting base participations with the exercise
         connectBaseParticipationsToExerciseAndSave(programmingExercise);
 
-        connectAuxiliaryRepositoriesToExercise(programmingExercise);
+        programmingExerciseRepository.saveAndFlush(programmingExercise);
```

Comment:
duplicated code (please extract a method, see comment below). And please explain shortly in inline code documentation, why saving early here is necessary.